### PR TITLE
FrontEnd: Event time and link type displayed when hovering a node

### PR DESCRIPTION
**Pull request by alrapal**
Original PR: ItJustWorksBetterTM/EiffelVis#32

# Display time and link type on hover

## Description:
* Modified lean-event to contain the link type 
* Use G6-graph's embedded edge labels to display link type 
* Use G6 graph's embedded node tooltips to display the time when the event was created. 
* We only change opacity to display or not display the labels since this is what impacts less the performance
  * On hover: 
    * Opacity of the link is increased (0.15 to 1)
    * Edge is made a bit thicker to better contrast
    * Opacity of the text is also increased (from 0 to 1) 
    * Label has shadow text to increase visibility (background colour is not usable in this case)
* We use the `groupByTypes: false` attribute in the graph option to allow controlling Z-index of the nodes. (So that labels would come on top of non hovered nodes). This is handled when a node is pushed into the graph to not impact performance. 
* Both outgoing and incoming edges are highlighted when hovering a node. 
 

## Useful links:
* [Z-index of elements in G6 graph
](https://antv-g6.gitee.io/en/docs/manual/middle/elements/methods/elementIndex)
* [Tooltips in G6 graph](https://g6.antv.vision/en/examples/tool/tooltip#tooltipPlugin )

### Overall App: 
Screenshot 2022-11-01 at 12 21 50

### When label on top of node
Screenshot 2022-11-01 at 12 22 42

### Outgoing and incoming edges 

Screenshot 2022-11-03 at 13 04 01
Screenshot 2022-11-03 at 13 04 07

## Testing:
* Tested locally with event sender using DB dump data. User experience was not impacted by these changes: no slow downs have been noticed.  

## Possible Improvements:
* We can adjust `fontSize` and `fill` to make things even more visible. 
* An alternative to the label would be to activate the tooltip embedded into edges but possible problem regarding performances and also difficulties to trigger them when hovering another item than themselves.

### CODE REVIEW

File: backend/tools/event_sender/src/main.rs [hunk 1]
```diff
@@ -82,7 +83,7 @@ async fn app() -> anyhow::Result<()> {
             .add_link(Link::new("Link0", true))
             .add_link(Link::new("Link1", true))
             .add_event(
-                Event::new("Event", "1.0.0")
+                Event::new(type_array[rand::thread_rng().gen_range(0..4)], "1.0.0")
```
Comment (RuthgerD) [inline backend/tools/event_sender/src/main.rs:86]: this conflicts with the event sender pr

File: frontend/.env.example [hunk 2]
```diff
@@ -0,0 +1,3 @@
+EIFFELVIS_URL = "@hostname:3001"
```
Comment (RuthgerD) [inline frontend/.env.example:1]: is this relevant to the PR?

File: frontend/src/components/G6Graph.svelte [hunk 3]
```diff
@@ -1,27 +1,27 @@
 <script lang="ts">
-    import { onMount } from "svelte";
```
Comment (RuthgerD) [inline frontend/src/components/G6Graph.svelte:2]: the auto format is a bit unfortunate, I know our config is not enforced atm but could you follow the style for now so that the diff is not ugly like this?

**Review by RuthgerD — COMMENTED** [review]
Looks good, minor nits for now; need to find time to actually try it out in person before giving a stamp of approval.

File: frontend/.env.example (hunk 2 above)
Comment (alrapal) [reply frontend/.env.example:1]: No! It slipped in during cherry picking, will change that! Thanks for pointing it out 😃

File: frontend/src/components/G6Graph.svelte (hunk 3 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:2]: Oh sorry, I did not think about this when formatting :x (stupid habits 😠)
I will notify the others about that!

File: backend/tools/event_sender/src/main.rs (hunk 1 above)
Comment (alrapal) [reply backend/tools/event_sender/src/main.rs:86]: That should not be there, good catch! I will revert this file to the latest commit on your master.

**Comment by alrapal** [comment]
> Looks good, minor nits for now; need to find time to actually try it out in person before giving a stamp of approval.

No problem! Better to test it out properly :D

File: backend/tools/event_sender/src/main.rs (hunk 1 above)
Comment (alrapal) [reply backend/tools/event_sender/src/main.rs]: It has been reverted to this commit [f8fd31a](https://github.com/ItJustWorksTM/EiffelVis/commit/f8fd31a77af249366ea7dd00540f928fb2690dd7)

File: frontend/src/components/G6Graph.svelte [hunk 4]
```diff
@@ -36,127 +36,245 @@
...
+  export const push = (ev: any) => {
+    ev.date = String(ev.time);
+    graph.addItem("node", ev, false, false);
+    for (const edge of ev.edges) {
+      graph.addItem("edge", {
+        source: ev.id,
+        target: edge.target,
+        label: edge.type,
+      }); // the type of link is connected to the label of the edge here.
+      edgesToBack(ev); // put all edges attached to the node behond the nodes (needed when using groupByTypes: false);
```
Comment (RuthgerD) [inline frontend/src/components/G6Graph.svelte:52]: this call should be outside the loop, now we are redrawing the graph for each edge

File: frontend/src/components/G6Graph.svelte [hunk 5]
```diff
@@ -36,127 +36,245 @@
...
+   * Helper method that will rearrange the order of items on the z-index (edges behind the nodes)
+   * To avoid iterating through the whole graph and update, we use this helper method inside the push method.
+   * This allows to only manipulate the nodes newly pushed into the graph.
+   * @param event containing a node and its edges.
+   */
+  const edgesToBack = (event) => {
+    const node = graph.findById(event.id);
+    if (node instanceof Node) {
+      const edges = node.getEdges();
+      edges.forEach((edge) => {
```
Comment (RuthgerD) [inline frontend/src/components/G6Graph.svelte:71]: A normal for loop would be more readable

File: frontend/src/components/G6Graph.svelte [hunk 6]
```diff
@@ -36,127 +36,245 @@
...
+    align-items: center;
+    border-color: #555555;
+    border-width: 1px;
+    box-shadow: rgb(35, 34, 34) 2px 2px 2px;
+    height: fit-content;
+    align-items: center;
+    width: fit-content;
+    font-size: 0.875rem;
+    color: #ffffff;
+    font-family: "ui-sans-serif", "system-ui", "-apple-system",
```
Comment (RuthgerD) [inline frontend/src/components/G6Graph.svelte:275]: this seems unneeded?

File: frontend/src/layout.ts [hunk 7]
```diff
@@ -71,4 +70,4 @@ export class StatefulLayout {
   getNodeColor() {
     return this.colors
   }
-}
+}
```
Comment (RuthgerD) [inline frontend/src/layout.ts:73]: new line

**Review by RuthgerD — CHANGES_REQUESTED** [review]
Would it be possible to keep the edges highlighted when a node is clicked?

current behavior when focused:
image

make it highlight the edges as well to give this effect:
image

File: frontend/src/components/G6Graph.svelte (hunk 4 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:52]: Will fix it!

File: frontend/src/components/G6Graph.svelte (hunk 5 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:71]: Quite subjective, I think the for.Each is more readable but I can change it to a normal loop if you wish so.

File: frontend/src/components/G6Graph.svelte (hunk 6 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:275]: Will look it up!

File: frontend/src/layout.ts (hunk 7 above)
Comment (alrapal) [reply frontend/src/layout.ts:73]: Not sure I understand what you mean here 😅

**Comment by alrapal** [comment]
> Would it be possible to keep the edges highlighted when a node is clicked?

It is possible, just out of scope of the feature as we defined it. But we can chat and see if we extend it on this one eventually or if we create a new issue for this  =).

File: frontend/src/components/G6Graph.svelte (hunk 4 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:52]: Fixed for the next commit (waiting for answers to questions before pushing to have one commit addressing all feedback)

File: frontend/src/components/G6Graph.svelte (hunk 5 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:71]: Fixed for the next commit (waiting for answers to questions before pushing to have one commit addressing all feedback)

File: frontend/src/components/G6Graph.svelte (hunk 6 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:275]: Fixed for the next commit (waiting for answers to questions before pushing to have one commit addressing all feedback)

File: frontend/src/layout.ts (hunk 7 above)
Comment (RuthgerD) [reply frontend/src/layout.ts:73]: files should always end with an empty new line, github shows a little cross symbol if it is not there.

**Comment by RuthgerD** [comment]
> > Would it be possible to keep the edges highlighted when a node is clicked?
> 
> It is possible, just out of scope of the feature as we defined it. But we can chat and see if we extend it on this one eventually or if we create a new issue for this =).
> 
> (Or at least discuss with you how you want it to be implemented since it can involve reworking the current listeners and more or less refactoring)

Its ok if its not included now, but I would like to see it for completeness; just make sure the code allows for it in the future.

File: frontend/src/components/G6Graph.svelte [hunk 8]
```diff
@@ -36,127 +36,256 @@
...
+  const edgesToBack = (event) => {
+    const node = graph.findById(event.id);
+    if (node instanceof Node) {
+      const edges = node.getEdges();
 
-        graph = new G6.Graph({
-            ...options,
-            container,
-        });
+      for(let i = 0; i < edges.length; i++){
```
Comment (RuthgerD) [inline frontend/src/components/G6Graph.svelte:73]: I mean a ranged for, I get why you were reluctant to use this but javascript has a modern range for:
```js
for (const edge of edges) {}
```

File: frontend/src/components/G6Graph.svelte (hunk 6 above)
Comment (RuthgerD) [reply frontend/src/components/G6Graph.svelte:286]: So all the fonts are needed?

File: frontend/src/components/G6Graph.svelte (hunk 8 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:73]: Aaah! That makes more sense haha! Will do!

Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:73]: This is how the code would look like in that case 

```
      for(const edge in edges){
        edges[edge].toBack();
      }
```

The problem with using `for ... in` is that the `const` is of type string and we need to use it as index value to make it work (which does not really add readability. 
But we can use `for ... of` as follow for more clarity? 

```
      for(const edge of edges){
        edge.toBack();
      }
```

Let me know which one you prefer =) (I would suggest the second one `for ... of`

Comment (RuthgerD) [reply frontend/src/components/G6Graph.svelte:73]: youre confusing me but just do the one that looks nice

Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:73]: Haha, sorry, I'll go for the second one then :D

File: frontend/src/components/G6Graph.svelte (hunk 6 above)
Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:286]: I am not sure about the fonts. I will ask around what was the intention with them =) 
I tested without and I don't see any impact on removing them but since it is not my contribution, I will ask in case there was something I am missing (like some fallback or something).

Comment (alrapal) [reply frontend/src/components/G6Graph.svelte:286]: @kam56 Do you have insights regarding the fonts? Can they be removed?

Comment (alrapal) [reply frontend/src/components/G6Graph.svelte]: I talked to @kam56 and based on the reason she gave me, I removed the fonts.

**Review by RuthgerD — APPROVED** [review]

**Comment by RuthgerD** [comment]
Thanks!
